File: pkg/executor/join/tagged_ptr.go
// Copyright 2024 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package join

import (
	"math/bits"
	"unsafe"
)

const maxTaggedBits = int8(24)
const maxTaggedMask = uint64(0xffffffffff)

// taggedPtr is a struct to save unsafe.Pointer with tagged value
// the value of unsafe.Pointer is actually an uint64 and in most
// cases, the n MSB in unsafe.Pointer is all zeros, which means we
// can save information in the n-MSB of a unsafe.Pointer.
// for example, if the 16 MSB in unsafe.Pointer is all zeros,
// then the value of unsafe.Pointer is like 0x0000xxxxxxxxxxxx
// we can save up to 16 bit information inside the unsafe.Pointer
// Assuming the tagged value is 0x1010, then the taggedPtr will
// be 0x1010xxxxxxxxxxxx
// However, we can not save the tagged value into unsafe.Pointer
// directly since go gc will check the validation of unsafe.Pointer,
// so in implementation, we save both the tagged value and the unsafe.Pointer
// into an uintptr.
type taggedPtr uintptr

type tagPtrHelper struct {
	// mask to get the tagged value in taggedPtr
	taggedMask uint64
}

func (th *tagPtrHelper) init(taggedBits uint8) {
	hashValueTaggedMask := uint64(1<<taggedBits) - 1
	hashValueTaggedOffset := 64 - taggedBits
	th.taggedMask = hashValueTaggedMask << hashValueTaggedOffset
}

func (th *tagPtrHelper) getTaggedValue(hashValue uint64) uint64 {
	return hashValue & th.taggedMask
}

func (*tagPtrHelper) toTaggedPtr(taggedValue uint64, ptr unsafe.Pointer) taggedPtr {
	ret := taggedPtr(0)
	// first save ptr into taggedPtr
	*(*unsafe.Pointer)(unsafe.Pointer(&ret)) = ptr
	// tag the value
	ret = taggedPtr(uint64(ret) | taggedValue)
	return ret
}

func (th *tagPtrHelper) toUnsafePointer(tPtr taggedPtr) unsafe.Pointer {
	// clear the tagged value
	tPtr = taggedPtr(uint64(tPtr) & ^th.taggedMask)
	return *(*unsafe.Pointer)(unsafe.Pointer(&tPtr))
}

func getTaggedBitsFromUintptr(ptr uintptr) uint8 {
	if sizeOfUintptr != 8 {
		return 0
	}
	// count leading zeros to determine the number of tagged bits
	return uint8(min(int8(bits.LeadingZeros64(uint64(ptr))), maxTaggedBits))
}

File: rust/crates/tidb-exec/src/tagged_ptr.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Top-bit pointer tagging metadata from `pkg/executor/join/tagged_ptr.go`.
//!
//! The Go source stores a raw pointer and a tag in one `uintptr`. This leaf
//! carries the same bit contract over an opaque raw address (`usize`) without
//! creating or dereferencing an unsafe pointer. Join hash-table ownership,
//! allocation, and concurrent access remain outside this value boundary.

/// Maximum number of high bits reserved for a tag in a 64-bit address.
pub const MAX_TAGGED_BITS: u8 = 24;
/// Lower 40 bits that are not reserved for a tag at the maximum width.
pub const MAX_TAGGED_MASK: u64 = 0xffffffffff;

/// An opaque raw address with optional high-bit tag bits set.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct TaggedPtr(usize);

impl TaggedPtr {
    /// Returns the encoded raw address.
    #[must_use]
    pub const fn raw(self) -> usize {
        self.0
    }
}

/// Mask helper for extracting and clearing high-bit tags.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct TagPtrHelper {
    tagged_mask: u64,
}

impl TagPtrHelper {
    /// Initializes the high-bit mask for a tag of `tagged_bits` width.
    pub fn init(&mut self, tagged_bits: u8) {
        if tagged_bits == 0 {
            self.tagged_mask = 0;
            return;
        }
        let hash_value_tagged_mask = (1_u64 << tagged_bits) - 1;
        let hash_value_tagged_offset = 64 - tagged_bits;
        self.tagged_mask = hash_value_tagged_mask << hash_value_tagged_offset;
    }

    /// Returns the configured high-bit tag portion of a raw value.
    #[must_use]
    pub fn get_tagged_value(&self, hash_value: u64) -> u64 {
        hash_value & self.tagged_mask
    }

    /// Returns the configured mask, useful for source-contract inspection.
    #[must_use]
    pub const fn tagged_mask(&self) -> u64 {
        self.tagged_mask
    }

    /// Encodes a tag into the high bits of a raw address.
    #[must_use]
    pub fn to_tagged_ptr(&self, tagged_value: u64, pointer: usize) -> TaggedPtr {
        TaggedPtr(pointer | tagged_value as usize)
    }

    /// Clears the tag bits and returns the original raw address.
    #[must_use]
    pub fn to_raw_pointer(&self, tagged_ptr: TaggedPtr) -> usize {
        TaggedPtr(tagged_ptr.0 & !(self.tagged_mask as usize)).raw()
    }
}

/// Determines how many leading zero bits can carry a tag on this address.
///
/// The source intentionally disables tagging on non-64-bit pointers and caps
/// the result at [`MAX_TAGGED_BITS`].
#[must_use]
pub fn get_tagged_bits_from_usize(pointer: usize) -> u8 {
    if usize::BITS != 64 {
        return 0;
    }
    (pointer.leading_zeros() as u8).min(MAX_TAGGED_BITS)
}
